app/scripts/lib/rpc-method-middleware/handlers/wallet-requestPermissions.ts

```diff
@@ -0,0 +1,165 @@
+  JsonRpcEngineEndCallback,
+} from '@metamask/json-rpc-engine';
+import {
+  CaveatTypes,
+  RestrictedMethods,
+} from '../../../../../shared/constants/permissions';
+import { PermissionNames } from '../../../controllers/permissions';
+
+export const requestPermissionsHandler = {
+  methodNames: [MethodNames.RequestPermissions],
+  implementation: requestPermissionsImplementation,
+  hookNames: {
+    getAccounts: true,
+    requestPermissionsForOrigin: true,
+    requestCaip25PermissionForOrigin: true,
+  },
+};
+
+type AbstractPermissionController = PermissionController<
+  PermissionSpecificationConstraint,
+  CaveatSpecificationConstraint
+>;
+
+type GrantedPermissions = Awaited<
+  ReturnType<AbstractPermissionController['requestPermissions']>
+>[0];
+
+/**
+ * Request Permissions implementation to be used in JsonRpcEngine middleware.
+ *
+ * @param req - The JsonRpcEngine request
+ * @param res - The JsonRpcEngine result object
+ * @param _next - JsonRpcEngine next() callback - unused
+ * @param end - JsonRpcEngine end() callback
+ * @param options - Method hooks passed to the method implementation
+ * @param options.getAccounts
+ * @param options.requestCaip25PermissionForOrigin
+ * @param options.requestPermissionsForOrigin
+ * @returns A promise that resolves to nothing
+ */
+async function requestPermissionsImplementation(
+  req: JsonRpcRequest<[RequestedPermissions]> & { origin: string },
+  res: PendingJsonRpcResponse<Json>,
+  _next: AsyncJsonRpcEngineNextCallback,
+  end: JsonRpcEngineEndCallback,
+  {
+    getAccounts,
+    requestPermissionsForOrigin,
+    requestCaip25PermissionForOrigin,
+  }: {
+    getAccounts: () => string[];
+    requestPermissionsForOrigin: (
+      requestedPermissions: RequestedPermissions,
+    ) => Promise<[GrantedPermissions]>;
+    requestCaip25PermissionForOrigin: (
+      requestedPermissions?: RequestedPermissions,
+    ) => Promise<
+      ValidPermission<
+        typeof Caip25EndowmentPermissionName,
+        Caveat<typeof Caip25CaveatType, Caip25CaveatValue>
+      >
+    >;
+  },
+) {
+  const { params } = req;
+
+  if (!Array.isArray(params) || !isPlainObject(params[0])) {
+    return end(invalidParams({ data: { request: req } }));
+  }
+
+  const [requestedPermissions] = params;
+  const legacyRequestedPermissions: Partial<
+    Pick<RequestedPermissions, 'eth_accounts' | 'endowment:permitted-chains'>
+  > = pick(requestedPermissions, [
+    RestrictedMethods.eth_accounts,
+    PermissionNames.permittedChains,
+  ]);
+  delete requestedPermissions[RestrictedMethods.eth_accounts];
+  delete requestedPermissions[PermissionNames.permittedChains];
+
+  const hasExpectedPermissions = Object.keys(requestedPermissions).length > 0;
+  const hasUnexpectedRequestedPermissions =
+    Object.keys(legacyRequestedPermissions).length > 0;
+
+  let caip25Endowment;
+  let caip25CaveatValue;
+  try {
+    if (hasExpectedPermissions || !hasUnexpectedRequestedPermissions) {
```

Comment:
I find the branching and logic in this handler confusing. What about potential requests to `wallet_requestPermissions` that don't need the CAIP-25 endowment?

This is not currently a blocker for Snaps, but eventually we want to expose APIs to do this.

Reply:
Thank you for pointing out this use case I wasn't aware of. I've fixed that here https://github.com/MetaMask/metamask-extension/pull/27847/commits/a8f4b55fb3a3fb84d14ce10d543e527f21b953ee